# SCM HW3 Q3b Simulation Code
This notebook simulates the Expected Marginal Seat Revenue (EMSR-b) model to calculate the expected revenue for a company managing limited capacity across multiple fare classes of airlines sets (1,2,3,4)

In Week 9, I learned that EMSR-b is a heuristic used in revenue management to protect capacity for higher-paying customers, by using the normal distribution of demand across fare classes.

## Thought Process

Therefore, I used Python to:

1. **Generate random demand** for each class using a normal distribution with the given mean and standard deviation.  

2. **Progressively allocate capacity** starting from the lowest fare class (Class 4) up to the highest (Class 1), following EMSR-b logic and using pre-calculated booking limits.  
3. **Calculate total revenue** from each simulation run based on how many seats were sold in each class and their respective fares.  
4. **Repeat the simulation 100,000 times** to capture demand variability and get a realistic average revenue.  
5. **Compute the expected revenue** by averaging the results across all runs.

This helped me understand that EMSR-b doesn’t just use average demand — it accounts for uncertainty and makes smarter capacity protection decisions. Running the simulation showed how demand variability affects revenue, and why fixed averages alone can be misleading.


In [1]:
import numpy as np

# Set random seed for reproducibility
np.random.seed(42)

# Fare prices per class
fares = np.array([3650, 2900, 1150, 600])

# Mean demand per class
mu = np.array([25, 35, 110, 70])

# Standard deviation per class
sigma = np.array([8, 16, 15, 10])

# Booking limits for each class (EMSR-b)
booking_limits = np.array([18, 49, 114, 119])

# Total seat capacity
capacity = 300

# Number of simulation runs
n_runs = 100_000

revenues = []

for _ in range(n_runs):
    # Step 1: Generate random demand from normal distribution
    demand = np.random.normal(mu, sigma)
    demand = np.maximum(0, demand)  # clip negative demand to 0

    # Step 2: Progressively compute x_4 to x_1
    x = np.zeros(4)  # seats sold per class
    remaining_capacity = capacity

    for i in reversed(range(4)):  # Start from class 4 to class 1
        x[i] = min(demand[i], booking_limits[i], remaining_capacity)
        remaining_capacity -= x[i]

    # Step 3: Compute total revenue for this run
    revenue = np.sum(x * fares)
    revenues.append(revenue)

# Step 5: Compute average (expected) revenue
expected_revenue = np.mean(revenues)

print(f"Expected Revenue over {n_runs:,} simulations: ${expected_revenue:,.2f}")


Expected Revenue over 100,000 simulations: $322,961.44


### Why the Expected Revenue is $322,961.44

The value **$322,961.44** is the result of simulating 100,000 possible demand scenarios using the EMSR-b booking limits.

Instead of assuming that demand is fixed and equal to the mean (like in a basic revenue calculation), the simulation treats demand as **random**, following a **normal distribution** with a given mean and standard deviation for each fare class.

Each run:
- Generates a different random demand outcome.
- Allocates available capacity starting from the lowest fare class (Class 4) to the highest (Class 1), respecting both the booking limits and remaining seats.
- Computes revenue based on how many seats are sold in each class.

By repeating this process 100,000 times and averaging the total revenue from all runs, we get a more **realistic estimate** of expected revenue when demand is uncertain.

The result — **$322,961.44** — reflects the fact that, due to variability in demand, we don't always fill all the booking limits, especially for higher-paying classes. It’s lower than the revenue calculated using just average demand (which was $335,700), because in real life, demand doesn’t always reach the mean.

This shows how EMSR-b accounts for **risk and uncertainty**, and why simulating is important when evaluating performance.
